File: java/src/main/java/br/com/rh/services/promotion/PromotionService.java
package br.com.rh.services.promotion;

import br.com.rh.enums.Role;
import br.com.rh.exceptions.PromotionException;
import br.com.rh.interfaces.TaxableReadjust;
import br.com.rh.models.Employee;

import java.math.BigDecimal;
import java.time.LocalDate;

public class PromotionService implements TaxableReadjust {
    private BigDecimal value;
    private LocalDate date;

    public PromotionService(BigDecimal value, LocalDate date) {
        this.value = value;
        this.date = date;
    }

    public void promote(Employee employee, boolean beatenMeta) {
        if (employee.getPersonalData().getRole() == Role.MANAGER) throw new PromotionException("Managers cannot be promoted");

        if(!beatenMeta) throw new PromotionException("Employee have not beaten the meta");

        Role newRole = employee.getPersonalData().getRole().getNextRole();
        employee.promote(newRole);
    }

    @Override
    public BigDecimal value() {
        return this.value;
    }

    @Override
    public LocalDate date() {
        return this.date;
    }

    @Override
    public BigDecimal taxValue() {
        return this.value.multiply(new BigDecimal("0.1"));
    }
}

File: typescript/src/services/promotion/promotion-service.ts
import { Role } from "../../enums/role.enum";
import { TaxableReadjust } from "../../interfaces/readjust";
import { Employee } from "../../models/employee";

export class PromotionService implements TaxableReadjust {
  private _value: number;
  private _date: Date;

  constructor(value: number, date: Date) {
    this._value = value;
    this._date = date;
  }

  promote(employee: Employee, isMetaBeaten: boolean) {
    if (employee.personalData.role === Role.MANAGER)
      throw new Error("Managers cannot be promoted");

    if (!isMetaBeaten) throw new Error("Employee have not beaten the meta");

    const newRole = this.getNextRole(employee.personalData.role);
    employee.promote(newRole);
  }

  taxValue(): number {
    return this._value * 0.2;
  }

  value(): number {
    return this._value * 2;
  }

  date(): Date {
    return this._date;
  }

  private getNextRole(actualRole: Role) {
    switch (actualRole) {
      case Role.ASSISTANT:
        return Role.ANALYST;
      case Role.ANALYST:
        return Role.SPECIALIST;
      case Role.SPECIALIST:
        return Role.MANAGER;
      default:
        throw new Error("Not a valid role");
    }
  }
}
